File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    #더하거나, 빼거나 ->2가지
    #numbers 배열의 길이가 5라면, 2^5이므로,
    # 길이가 n일때 2^n
    #적절히 숫자 조합하여, 그 합이 target이 되도록
    que = deque()
    calc_value = 0
    
    que.append((0,0))
    while que :
        total_sum, cur_idx = que.popleft()
        
        if cur_idx == len(numbers) :
            if total_sum == target :
                answer+=1
                
        else:
            que.append((total_sum + numbers[cur_idx], cur_idx+1))
            que.append((total_sum - numbers[cur_idx], cur_idx+1))
                
        
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) {
        int answer = 0;
        Queue <int[] > q = new ArrayDeque<>();
        
        q.offer (new int[] {0,0});
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            
            int idx = cur[0];
            int sum = cur[1];
            
            if (idx == numbers.length) {
                if (sum == target) {
                    answer +=1;
                }
                continue;
            }
            
            q.offer (new int [] {idx+1,sum + numbers[idx]});
            
            q.offer (new int [] {idx+1, sum -numbers[idx]});
            
            
        }
        
        return answer;
    }
}